packages/core/bootstrap/src/lib/util.ts

```diff
@@ -373,3 +373,101 @@ export const getRequiredEnvWithFallback = (
 
   throw new RequiredEnvError(getEnvName(primary, prefix))
 }
+
+//  URL Encoding
+
+const charsToEncode = {
+  ':': '%3A',
+  '/': '%2F',
+  '?': '%3F',
+  '#': '%23',
+  '[': '%5B',
+  ']': '%5D',
+  '@': '%40',
+  '!': '%21',
+  $: '%24',
+  '&': '%26',
+  "'": '%27',
+  '(': '%28',
+  ')': '%29',
+  '*': '%2A',
+  '+': '%2B',
+  ',': '%2C',
+  ';': '%3B',
+  '=': '%3D',
+  '%': '%25',
+  ' ': '%20',
+  '"': '%22',
+  '<': '%3C',
+  '>': '%3E',
+  '{': '%7B',
+  '}': '%7D',
+  '|': '%7C',
+  '^': '%5E',
+  '`': '%60',
+  '\\': '%5C',
+}
+
+/**
+ * Check whether the given string contains characters in the given whitelist.
+ * @param str The string to check.
+ * @param whitelist The string array of whitelist entries. Returns true if any of these are found in 'str', otherwise returns false.
+ * @returns {boolean}
+ */
+const stringHasWhitelist = (str: string, whitelist: string[]): boolean =>
+  whitelist.some((el) => str.includes(el))
+
+/**
+ * Manually iterate through a given string and replace unsafe/reserved characters with encoded values (unless a character is whitelisted)
+ * @param str The string to encode.
+ * @param whitelist The string array of whitelist entries.
+ * @returns {string}
+ */
+const percentEncodeString = (str: string, whitelist: string[]): string =>
+  str
+    .split('')
+    .map((char) => {
+      const encodedValue = charsToEncode[char as keyof typeof charsToEncode]
+      return encodedValue && !whitelist.includes(char) ? encodedValue : char
+    })
+    .join('')
+
+/**
+ * Build a URL path using the given pathTemplate and params. If a param is found in pathTemplate, it will be inserted there; otherwise, it will be ignored.
+ * eg.) pathTemplate = "/from/:from/to/:to" and params = { from: "ETH", to: "BTC", note: "hello" } will become "/from/ETH/to/BTC"
+ * @param pathTemplate The path template for the URL path. Each param to include in the path should have a prefix of ':'.
+ * @param params The object containing keys & values to be added to the URL path.
+ * @param whitelist The list of characters to whitelist for the URL path (if a param contains one of your whitelisted characters, it will not be encoded).
+ * @returns {string}
+ */
+export const buildUrlPath = (pathTemplate = '', params = {}, whitelist = ''): string => {
+  const allowedChars = whitelist.split('')
+
+  for (const param in params) {
+    const value = params[param as keyof typeof params]
+    if (!value) continue
+
+    // If string contains a whitelisted character: manually replace any non-whitelisted characters with percent encoded values. Otherwise, encode the string as usual.
+    const encodedValue = stringHasWhitelist(value, allowedChars)
+      ? percentEncodeString(value, allowedChars)
+      : encodeURIComponent(value)
+
+    if (pathTemplate.includes(`:${param}`))
+      pathTemplate = pathTemplate.replace(`:${param}`, encodedValue)
```

Comment:
I think you can move the `if` statement up to here (lmk if I'm missing something):
```suggestion
    const value = params[param as keyof typeof params]
    if (!value || !pathTemplate.includes(`:${param}`)) continue

    // If string contains a whitelisted character: manually replace any non-whitelisted characters with percent encoded values. Otherwise, encode the string as usual.
    const encodedValue = stringHasWhitelist(value, allowedChars)
      ? percentEncodeString(value, allowedChars)
      : encodeURIComponent(value)

      pathTemplate = pathTemplate.replace(`:${param}`, encodedValue)
```

Reply:
You're right about the `if (pathTemplate.includes(':${param}'))`, but actually, I think it can probably be removed altogether rather than being moved up. That `if` was needed for an earlier version of the method but it shouldn't be needed at this point, since string.replace shouldn't do anything if the param isn't found